### PR TITLE
[Bug] Fix TRTLLM Block FP8 MoE Monolithic

### PR DESCRIPTION
**Pull request by wzhao18**
## Purpose
Fix #36295

## Test Plan
```
vllm serve deepseek-ai/DeepSeek-R1 -tp 8
```
```
lm_eval --model local-completions --model_args "base_url=http://0.0.0.0:8000/v1/completions,max_length=8192,tokenized_requests=False,tokenizer_backend=None,num_concurrent=32" --tasks gsm8k --num_fewshot 5
```

## Test Result
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.953|±  |0.0058|
|     |       |strict-match    |     5|exact_match|↑  |0.953|±  |0.0058|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 1]
```diff
@@ -182,12 +182,12 @@ def _apply_per_block(
...
         assert activation == MoEActivation.SILU
 
-        if e_score_correction_bias is not None:
-            e_score_correction_bias = e_score_correction_bias.to(hidden_states.dtype)
-
         if self.routing_method_type == RoutingMethodType.DeepSeekV3:
             router_logits = router_logits.to(torch.float32)
 
+        if e_score_correction_bias is not None:
+            e_score_correction_bias = e_score_correction_bias.to(router_logits.dtype)
```
[the remark this answers is not in the document]
Comment (wzhao18) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py]: I am not sure which model triggers the code path. Won't be able to test and fix this.

**Review by bnellnm — APPROVED** [review]

**Review by pavanimajety — APPROVED** [review]
LGTM, can we add a note in the PR or comment in the code saying what determines the e_score_correction_bias datatype? It's typically based on the checkpoint and kernel checkpoint, so hardcoding with hidden_states or router_logits, could  work for a few models while not work for the others. 
For example deepseek uses -
```
model.layers.3.mlp.gate.e_score_correction_bias | [256] | F32
```

**Comment by wzhao18** [comment]
Talked to Pavani offline. I think we could merge this patch for now. but it may be worth checking whether we should instead make change to the kernel side to support the built-in dtype for `e_score_correction_bias` from the model checkpoint.

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py (hunk 1 above)
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py]: this works, but is a hack. Do we know what dtype is required by the kernel here? Is it just bf16?

**Review by ProExpertProg — APPROVED** [review]

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by robertgshaw2-redhat** [comment]
LGTM, thanks for the fix

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by robertgshaw2-redhat** [comment]
verified same accuracy with FP32 and Bf16
